numpy/core/src/umath/loops_minmax.dispatch.c.src

```diff
@@ -0,0 +1,671 @@
+/*@targets
+ ** $maxopt baseline
+ ** neon
+ **/
+#define _UMATHMODULE
+#define _MULTIARRAYMODULE
+#define NPY_NO_DEPRECATED_API NPY_API_VERSION
+
+#include "simd/simd.h"
+#include "loops_utils.h"
+#include "loops.h"
+#include "lowlevel_strided_loops.h"
+// Provides the various *_LOOP macros
+#include "fast_loop_macros.h"
+
+
+//##############################################################################
+//## Maximum, Minimum
+//##############################################################################
+
+#ifdef NPY_HAVE_NEON
+//******************************************************************************
+//** NEON Min / Max
+//******************************************************************************
+
+
+//------------------------------------------------------------------------------
+//-- Integer
+//------------------------------------------------------------------------------
+
+
+/**begin repeat
+ * #TYPE = BYTE, UBYTE, SHORT, USHORT, INT, UINT,
+ *         LONG, ULONG, LONGLONG, ULONGLONG#
+ * #type = npy_byte, npy_ubyte, npy_short, npy_ushort, npy_int, npy_uint,
+ *         npy_long, npy_ulong, npy_longlong, npy_ulonglong#
+ * #sfx = s8, u8, s16, u16, s32, u32,
+ *        s64, u64, s64, u64#
+ * #HAVE_NEON_IMPL = 1*8, HAVE_NEON_IMPL_LONGLONG*2#
+ */
+
+// Implementation below assumes longlong and ulonglong are 64-bit.
+#if @HAVE_NEON_IMPL@
+
+/**begin repeat1
+* Arithmetic
+* # kind = maximum, minimum#
+* # OP = >, <#
+* # vop = max, min#
+*/
+
+#define SCALAR_OP @TYPE@_scalar_@vop@
+static inline @type@ SCALAR_OP(@type@ a, @type@ b){
+    return ((a @OP@ b) ? a : b);
+}
+
+static inline npy_intp
+simd_reduce_@TYPE@_@kind@(char **args, npy_intp const *dimensions, npy_intp const *steps, npy_intp i)
+{
```

Comment:
Here is a fact, any ufunc's reduction has an identity, the value of the first element of the second input array. @seberg please correct me if I'm wrong. that's mean the input and output array can't be empty at least the length greater than zero.

In the light of the above, your SIMD kernel should use this identity value as a default value for the final accumulated vector, so we loop vector by vector to deal with small arrays or when we deal with large SIMD width > 128-bit. see the following example:

```C
// any ufunc reduce always has an identity
// which is the value of the first element of the second input array
assert(len > 0);
const int nlanes = npyv_nlanes_@sfx@;
npyv_@sfx@ acc = npyv_setall_@sfx@(op1[0]); // final accumulator
for (; len >= nlanes*8; len -= nlanes*8, ip += nlanes*8) {
    // unroll goes here
}
for (; len >= nlanes; len -= nlanes, ip += nlanes) {
    acc = npyv_@vop@_@sfx@(npyv_load_@sfx@(ip), acc);
}
npyv_lanetype_@sfx@ r = npyv_reduce_@vop@_@sfx@(acc);
for (; len > 0; --len, ++ip) {
    const npyv_lanetype_@sfx@ a = *ip;
    r = SCALAR_OP(r, a);
}
*op1 = r;
```

Reply:
Actually... no I do not think you can, reading the 0th element is not actually well defined currently, the macro checking for a "reduction" doesn't ensure that it actually *is* a reduction (which indeed would have a guaranteed 1-element), but only that it effectively is one.

This is super obscure though, and pretty much implausible to hit (if you are in the "reduce" branch).  So probably the fix may be that NumPy should guarantee to never call an inner-loop with `N == 0` (`*dimensions == 0`)?
(This might be one of the reason NumPy historically "over-allocates" empty arrays, but that habit is one of those 80% band-aid fixes, only...)

Reply:
thanks for the clarification, then we have to test empty arrays.